std/algorithm/sorting.d

```diff
@@ -1048,147 +1048,287 @@ unittest
+            this._lastBackIndex = backIndex;
+    }
 
-    void adjustPosition(uint candidate = 0)()
+    import std.functional : binaryFun;
+    private alias comp = binaryFun!less;
+
+    import std.typetuple : anySatisfy;
+    static if (anySatisfy!(isInfinite, Rs))
     {
-        static if (candidate == Rs.length)
+        enum bool empty = false; // propagate infiniteness
+    }
+    else
+    {
+        @property bool empty()
         {
-            _crt = _crt.max;
+            if (_lastFrontIndex == size_t.max)
+                return true;
+            static if (isBidirectional)
+                return _lastBackIndex == size_t.max;
+            else
+                return false;
         }
-        else
+    }
+
+    @property auto ref front()
+    {
+        final switch (_lastFrontIndex)
         {
-            if (_r[candidate].empty)
+            foreach (i, _; Rs)
             {
-                adjustPosition!(candidate + 1)();
-                return;
+            case i:
+                assert(!source[i].empty);
+                return source[i].front;
             }
-            foreach (i, U; Rs[candidate + 1 .. $])
+        }
+        assert(false);
+    }
+
+    private size_t frontIndex()
+    {
+        size_t bestIndex = size_t.max; // indicate undefined
+        Unqual!ElementType bestElement;
+        foreach (i, _; Rs)
+        {
+            if (!source[i].empty)
             {
-                enum j = candidate + i + 1;
-                if (_r[j].empty) continue;
-                if (comp(_r[j].front, _r[candidate].front))
+                if (bestIndex == size_t.max || // either this is the first or
+                    comp(source[i].front, bestElement))
                 {
-                    // a new candidate was found
-                    adjustPosition!(j)();
-                    return;
+                    bestIndex = i;
+                    bestElement = source[i].front;
                 }
             }
-            // Found a successful candidate
-            _crt = candidate;
         }
+        return bestIndex;
     }
 
-public:
-    alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
-    static assert(!is(CommonType!(staticMap!(.ElementType, Rs)) == void),
-        typeof(this).stringof ~ ": incompatible element types.");
-
-    ///
-    this(Rs rs)
+    void popFront()
     {
-        this._r = rs;
-        adjustPosition();
+        final switch (_lastFrontIndex)
+        {
+            foreach (i, _; Rs)
+            {
+            case i:
+                debug Unqual!ElementType previousFront = source[i].front();
+                source[i].popFront();
+                debug if (!source[i].empty)
+                {
+                    assert(previousFront == source[i].front ||
+                           comp(previousFront, source[i].front),
+                           "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                }
+                break;
```

Comment:
This breaks the foreach loop but not the switch. So you'd need a break for the switch as well.